scaleway/resource_instance_user_data.go

```diff
@@ -0,0 +1,197 @@
+package scaleway
+
+import (
+	"bytes"
+	"context"
+	"io"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/scaleway/scaleway-sdk-go/api/instance/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+func resourceScalewayInstanceUserData() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceScalewayInstanceUserDataCreate,
+		ReadContext:   resourceScalewayInstanceUserDataRead,
+		UpdateContext: resourceScalewayInstanceUserDataUpdate,
+		DeleteContext: resourceScalewayInstanceUserDataDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create:  schema.DefaultTimeout(defaultInstanceServerWaitTimeout),
+			Read:    schema.DefaultTimeout(defaultInstanceServerWaitTimeout),
+			Update:  schema.DefaultTimeout(defaultInstanceServerWaitTimeout),
+			Delete:  schema.DefaultTimeout(defaultInstanceServerWaitTimeout),
+			Default: schema.DefaultTimeout(defaultInstanceServerWaitTimeout),
+		},
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"server_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "The ID of the server",
+				ValidateFunc: validationUUIDWithLocality(),
+			},
+			"key": {
```

Comment:
Why not use an HCL maps: https://www.terraform.io/language/expressions/types#maps-objects ?

Reply:
Taking this approach we allow the client to update key by key instead update all the user-data.